allensdk/brain_observatory/behavior/behavior_project_lims_api.py

```diff
@@ -0,0 +1,300 @@
+
+
+class BehaviorProjectLimsApi(BehaviorProjectBase):
+    def __init__(self, postgres_engine, app_engine):
+        """ Downloads visual behavior data from the Allen Institute's
+        internal Laboratory Information Management System (LIMS). Only
+        functional if connected to the Allen Institute Network. Used to load
+        data into BehaviorProjectCache.
+
+        Typically want to construct an instance of this class by calling
+            `BehaviorProjectLimsApi.default()`.
+
+        Note -- Currently the app engine is unused because we aren't yet
+        supporting the download of stimulus templates for visual behavior
+        data. This feature will be added at a later date.
+
+        Parameters
+        ----------
+        postgres_engine :
+            used for making queries against the LIMS postgres database. Must
+            implement:
+                select : takes a postgres query as a string. Returns a pandas
+                    dataframe of results
+                fetchall : takes a postgres query as a string. If there is
+                    exactly one column in the response, return the values as a
+                    list.
+        app_engine :
+            used for making queries agains the lims web application. Must
+            implement:
+                stream : takes a url as a string. Returns an iterable yielding
+                the response body as bytes.
+        """
+        self.postgres_engine = postgres_engine
+        self.app_engine = app_engine
+        self.logger = logging.getLogger("BehaviorProjectLimsApi")
+
+    @classmethod
+    def default(
+        cls,
+        pg_kwargs: Optional[Dict[str, Any]] = None,
+        app_kwargs: Optional[Dict[str, Any]] = None) -> \
+            "BehaviorProjectLimsApi":
+        """Construct a BehaviorProjectLimsApi instance with default
+        postgres and app engines.
+
+        :param pg_kwargs: dict of keyword arguments to pass to the
+        PostgresQueryMixin class instance. Valid arguments include:
+            "dbname", "user", "host", "password", "port". Will use
+            defaults in PostGresQueryMixin.__init__ if unspecified.
+        :type pg_kwargs: dict
+        :param app_kwargs: dict of keyword arguments to pass to the
+            HTTPEngine class instance. Valid arguments include:
+            "scheme", "host". Will default to scheme=http, host=lims2
+            if left unspecified.
+        :type app_kwargs: dict
+        :rtype: BehaviorProjectLimsApi
+        """
+        _pg_kwargs = pg_kwargs or dict()
+
+        _app_kwargs = {"scheme": "http", "host": "lims2"}
+        if app_kwargs:
+            _app_kwargs.update(app_kwargs)
+
+        pg_engine = PostgresQueryMixin(**_pg_kwargs)
+        app_engine = HttpEngine(**_app_kwargs)
+        return cls(pg_engine, app_engine)
+
+    @staticmethod
+    def _build_in_list_selector_query(
+            col,
+            valid_list: Optional[List[Union[str, int]]] = None,
```

Comment:
include float here too?

Reply:
technically the type is anything that implements __str__. So truly it would have to be an AbstractBaseClass with __str__, say called `SupportsStr`, and the type would be `SupportsStr[T]`.

Reply:
Updated